File: Reaktoro/Core/Material.cpp
// Reaktoro is a unified framework for modeling chemically reactive systems.
//
// Copyright © 2014-2022 Allan Leal
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this library. If not, see <http://www.gnu.org/licenses/>.

#include "Material.hpp"

// Reaktoro includes
#include <Reaktoro/Common/Algorithms.hpp>
#include <Reaktoro/Common/ParseUtils.hpp>

namespace Reaktoro {

struct Material::Impl
{
    /// The chemical formula of the substance (e.g., `HCO3-`).
    String formula;

    /// The element symbols and their coefficients (e.g., `{{"H", 1}, {"C", 1}, {"O", 3}}` for `HCO3-`).
    Pairs<String, double> elements;

    /// The electric charge in the chemical formula (e.g., `-1` for `HCO3-`).
    double charge = {};

    /// Construct an object of type Impl.
    Impl()
    {}

    /// Construct an object of type Impl with given formula.
    Impl(String formula)
    : formula(formula), elements(parseChemicalFormula(formula)), charge(parseElectricCharge(formula))
    {}

    /// Construct an object of type Impl with given data.
    Impl(String formula, Pairs<String, double> elements, double charge)
    : formula(formula), elements(elements), charge(charge)
    {}

    /// Return the symbols of the elements.
    auto symbols() const -> Strings
    {
        return vectorize(elements, RKT_LAMBDA(pair, pair.first));
    }

    /// Return the coefficients of the elements.
    auto coefficients() const -> Vec<double>
    {
        return vectorize(elements, RKT_LAMBDA(pair, pair.second));
    }

    /// Return the coefficient of an element symbol in the chemical formula.
    auto coefficient(const String& symbol) const -> double
    {
        for(auto const& element : elements)
            if(element.first == symbol)
                return element.second;
        return 0.0;
    }
};

Material::Material()
: pimpl(new Impl())
{}

Material::Material(const char* formula)
: Material(String(formula))
{}

Material::Material(String formula)
: pimpl(new Impl(formula))
{}

Material::Material(String formula, Pairs<String, double> symbols, double charge)
: pimpl(new Impl(formula, symbols, charge))
{}

auto Material::str() const -> const String&
{
    return pimpl->formula;
}

auto Material::elements() const -> const Pairs<String, double>&
{
    return pimpl->elements;
}

auto Material::symbols() const -> Strings
{
    return pimpl->symbols();
}

auto Material::coefficients() const -> Vec<double>
{
    return pimpl->coefficients();
}

auto Material::coefficient(const String& symbol) const -> double
{
    return pimpl->coefficient(symbol);
}

auto Material::charge() const -> double
{
    return pimpl->charge;
}

auto Material::equivalent(const Material& other) const -> bool
{
    return elements().size() == other.elements().size() &&
        contained(elements(), other.elements()) &&
        charge() == other.charge();
}

auto Material::equivalent(const Material& f1, const Material& f2) -> bool
{
    return f1.equivalent(f2);
}

Material::operator String() const
{
    return str();
}

Material::operator Pairs<String, double>() const
{
    return pimpl->elements;
}

auto operator<(const Material& lhs, const Material& rhs) -> bool
{
    return lhs.str() < rhs.str();
}

auto operator==(const Material& lhs, const Material& rhs) -> bool
{
    return lhs.str() == rhs.str();
}

} // namespace Reaktoro

File: Reaktoro/Utils/Material.py
# Reaktoro is a unified framework for modeling chemically reactive systems.
#
# Copyright © 2014-2022 Allan Leal
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library. If not, see <http://www.gnu.org/licenses/>.


from reaktoro import *
import pytest


def testMaterial():

    db = SupcrtDatabase("supcrtbl")

    solution = AqueousPhase(speciate("H O Na Cl Ca C Mg Si"), exclude("organic"))
    solution.setActivityModel(chain(ActivityModelHKF(), ActivityModelDrummond("CO2")))

    gases = GaseousPhase("H2O(g) CO2(g) O2(g) CH4(g)")
    gases.setActivityModel(ActivityModelPengRobinson())

    minerals = MineralPhases("Halite Calcite Magnesite Dolomite Quartz")

    system = ChemicalSystem(db, solution, gases, minerals)

    brine = Material(system)
    brine.add("H2O"    , 0.30, "kg")  # using formula
    brine.add("H2O(aq)", 0.70, "kg")  # using species name
    brine.add("NaCl"   , 1.00, "mol") # using formula
    brine.add("CaCl2"  , 0.10, "mol") # using formula
    brine.add("MgCl2"  , 0.05, "mol") # using formula

    rock = Material(system)
    rock.add("CaCO3", 1.00, "mol")    # using formula
    rock.add("SiO2" , 1.00, "mol")    # using formula

    mix = brine(1.0, "kg") + rock(1.0, "kg")

    bmix = mix.componentAmounts()  # the amounts of elements/charge in mix material

    state = ChemicalState(system)

    state = mix.equilibrate()

    assert mix.result().optima.succeeded
    assert mix.result().optima.iterations == 48

    assert state.temperature() == pytest.approx(25.0 + 273.15)
    assert state.pressure() == pytest.approx(1.0 * 1e5)
    assert state.componentAmounts().asarray() == pytest.approx(bmix)  # ensure computed state has element/charge amounts equal to those in mix material

    state = mix.equilibrate(60.0, "celsius", 10.0, "bar")

    assert mix.result().optima.succeeded
    assert mix.result().optima.iterations == 47

    assert state.temperature() == pytest.approx(60.0 + 273.15)
    assert state.pressure() == pytest.approx(10.0 * 1e5)
    assert state.componentAmounts().asarray() == pytest.approx(bmix)  # ensure computed state has element/charge amounts equal to those in mix material

